cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+    fn new(len: usize) -> Self {
+        let mut nodes = Vec::with_capacity(len);
+        nodes.resize_with(len, || Node::new(len));
+        Self { nodes }
+    }
+
+    /// True when there are no edges in the graph.
+    fn is_empty(&self) -> bool {
+        self.nodes.iter().all(|node| node.degree == 0)
+    }
+
+    /// Condense the overlap information down into individual errors.
+    fn report(&mut self, env: &Env) -> Vec<Error> {
+        let mut rules: Vec<usize> = self
+            .nodes
+            .iter()
+            .enumerate()
+            .filter_map(|(id, node)| if node.degree == 0 { None } else { Some(id) })
+            .collect();
+
+        rules.sort_by_cached_key(|id| self.nodes[*id].degree);
+
+        let mut errors = Vec::new();
+
+        let get_info = |id| {
+            let rule = env.get_rule(id);
+
+            let src = env.get_source(rule.pos.file);
+            let span = Span::new_single(rule.pos);
+            (src, span)
+        };
+
+        // Work backwards through the ids to find the nodes with the largest conflict first.
+        for id in rules.into_iter().rev() {
+            let node = self.remove_edges(RuleId(id));
+            if node.degree == 0 {
+                continue;
+            }
+
+            // build the real error
+            let mut rules = vec![get_info(RuleId(id))];
+
+            rules.extend(
+                node.edges
+                    .into_iter()
+                    .enumerate()
+                    .filter_map(|(ix, present)| {
+                        if present {
+                            Some(get_info(RuleId(ix)))
+                        } else {
+                            None
+                        }
+                    }),
+            );
+
+            errors.push(Error::OverlapError {
+                msg: String::from("rules are overlapping"),
+                rules,
+            });
+        }
+
+        errors
+    }
+
+    /// Remove all the edges for this rule in the graph, returning the original `Node` contents for
+    /// further processing.
+    fn remove_edges(&mut self, id: RuleId) -> Node {
+        let mut node = Node::new(self.nodes.len());
+        std::mem::swap(&mut self.nodes[id.0], &mut node);
+
+        for (ix, other) in node.edges.iter().copied().enumerate() {
+            if other {
+                self.nodes[ix].remove_edge(id);
+            }
+        }
+
+        node
+    }
+
+    /// Add a bidirectional edge between two rules in the graph.
+    fn add_edge(&mut self, a: RuleId, b: RuleId) {
+        // edges are undirected
+        self.nodes[a.0].add_edge(b);
+        self.nodes[b.0].add_edge(a);
+    }
+
+    /// Register all of the rules in the matrix as overlapping.
+    fn overlap_error(&mut self, matrix: Matrix) {
+        for (ix, rule) in matrix.rows.iter().enumerate() {
+            for other in &matrix.rows[ix + 1..] {
+                self.add_edge(rule.rule, other.rule);
+            }
+        }
```

Comment:
I don't think there's a good idiom for "loop over all pairs in this slice", but I found this version of it confusing for a moment. Maybe this instead?

```suggestion
        let mut rows = &matrix.rows[..];
        while let Some((rule, rest)) = rows.split_first() {
            for other in rest {
                self.add_edge(rule.rule, other.rule);
            }
            rows = others;
        }
```